airflow/providers/amazon/aws/executors/ecs/utils.py

```diff
@@ -0,0 +1,272 @@
+
+    def update_task(self, task: EcsExecutorTask):
+        """Updates the state of the given task based on task ARN."""
+        self.tasks[task.task_arn] = task
+
+    def task_by_key(self, task_key: TaskInstanceKey) -> EcsExecutorTask:
+        """Get a task by Airflow Instance Key."""
+        arn = self.key_to_arn[task_key]
+        return self.task_by_arn(arn)
+
+    def task_by_arn(self, arn) -> EcsExecutorTask:
+        """Get a task by AWS ARN."""
+        return self.tasks[arn]
+
+    def pop_by_key(self, task_key: TaskInstanceKey) -> EcsExecutorTask:
+        """Deletes task from collection based off of Airflow Task Instance Key."""
+        arn = self.key_to_arn[task_key]
+        task = self.tasks[arn]
+        del self.key_to_arn[task_key]
+        del self.key_to_task_info[task_key]
+        del self.arn_to_key[arn]
+        del self.tasks[arn]
+        if task_key in self.key_to_failure_counts:
+            del self.key_to_failure_counts[task_key]
+        return task
+
+    def get_all_arns(self) -> list[str]:
+        """Get all AWS ARNs in collection."""
+        return list(self.key_to_arn.values())
+
+    def get_all_task_keys(self) -> list[TaskInstanceKey]:
+        """Get all Airflow Task Keys in collection."""
+        return list(self.key_to_arn.keys())
+
+    def failure_count_by_key(self, task_key: TaskInstanceKey) -> int:
+        """Get the number of times a task has failed given an Airflow Task Key."""
+        return self.key_to_failure_counts[task_key]
+
+    def increment_failure_count(self, task_key: TaskInstanceKey):
+        """Increment the failure counter given an Airflow Task Key."""
+        self.key_to_failure_counts[task_key] += 1
+
+    def info_by_key(self, task_key: TaskInstanceKey) -> EcsTaskInfo:
+        """Get the Airflow Command given an Airflow task key."""
+        return self.key_to_task_info[task_key]
+
+    def __getitem__(self, value):
+        """Gets a task by AWS ARN."""
+        return self.task_by_arn(value)
+
+    def __len__(self):
+        """Determines the number of tasks in collection."""
+        return len(self.tasks)
+
+
+def _recursive_flatten_dict(nested_dict):
+    """
+    Recursively unpack a nested dict and return it as a flat dict.
+
+    For example, _flatten_dict({'a': 'a', 'b': 'b', 'c': {'d': 'd'}}) returns {'a': 'a', 'b': 'b', 'd': 'd'}.
+    """
+    items = []
+    for key, value in nested_dict.items():
+        if isinstance(value, dict):
+            items.extend(_recursive_flatten_dict(value).items())
+        else:
+            items.append((key, value))
+    return dict(items)
+
+
+def parse_assign_public_ip(assign_public_ip):
+    """Convert "assign_public_ip" from True/False to ENABLE/DISABLE."""
+    return "ENABLED" if assign_public_ip == "True" else "DISABLED"
+
+
+def snake_to_camel(_key):
+    split_key = _key.split("_")
+    first_word = split_key[0]
+    return first_word[0].lower() + first_word[1:] + "".join(word.title() for word in split_key[1:])
```

Comment:
Same as [inflection.camelize](https://inflection.readthedocs.io/en/latest/#inflection.camelize).
We already have this really nice library in constraints: https://github.com/apache/airflow/blob/8fa683885f7043ae00a7f7a1cd69202191f0e8d9/constraints-3.8.txt#L371, all we need it add as dependency to Amazon Provider and use it instead

Reply:
I've been thinking about this one.   I'm not against the idea, but is there some way we can add it as a dependency only if the executor is being used?  I feel like it's overkill and an undue addition to make it a dependency for the package.   I ran into a lot of friction when I added otel as an Airflow dependency even if otel wasn't being used for metrics... I think if we can think of a more elegant way to do it so it was only required and installed if an AWS Executor was being used, that would be a cleaner user experience.

Reply:
Nevermind, I'll add it should be up on Monday

Reply:
This is addressed in https://github.com/apache/airflow/pull/34381/commits/40f0458e65e475bcfe430b4a36efd44bc2d28432, thanks for the suggestion.